### PR TITLE
Cluster: Only create persistent warning for cluster time skew when cluster DB is operational

### PR DESCRIPTION
**Pull request by tomponline**
- Previously this was preventing the heartbeat request from being handled normally if the cluster DB was fully functioning as the query to retrieve the local name failed.
-  Avoids the DB query to look up the name in normal situations and only when a skew is detected or fixed do we need to look up the name.
- Also cleans up the logging to log the actual error.
- Still log the warning to syslog though as that doesn't require DB access.
- Also fixes bug in warnings system that was causing a previously resolved (but not yet deleted) warning from being reinstated as "new" (status change for existing warning) if the warning reoccurred. We now set the status back to new if the status is resolved (as opposed to acknowledged).

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>

### CODE REVIEW

**Comment by tomponline** [comment]
@monstermunchkin thinking about this more, is there any reason that `g.Cluster.UpsertWarning` shouldn't just not have a nodeName argument, and instead it would look up the local node name itself (and return an error if it couldn't)?

**Comment by monstermunchkin** [comment]
> @monstermunchkin thinking about this more, is there any reason that `g.Cluster.UpsertWarning` shouldn't just not have a nodeName argument, and instead it would look up the local node name itself (and return an error if it couldn't)?

Nope. I believe the spec allows node name to be empty, so returning an error would be incorrect. IMHO the caller should decide whether a node name is needed/required.

**Comment by tomponline** [comment]
> > @monstermunchkin thinking about this more, is there any reason that `g.Cluster.UpsertWarning` shouldn't just not have a nodeName argument, and instead it would look up the local node name itself (and return an error if it couldn't)?
> 
> Nope. I believe the spec allows node name to be empty, so returning an error would be incorrect. IMHO the caller should decide whether a node name is needed/required.

My observation was that we are seeing quite a bit of code duplication for the common scenario of logging a warning for the local member (and in cases where the local name isn't available we probably don't want to log the warning under no name).

Perhaps a new helper function called `UpsertWarningNode()`?

**Comment by monstermunchkin** [comment]
> Perhaps a new helper function called UpsertWarningNode()?

We could do that.

**Comment by tomponline** [comment]
@monstermunchkin thanks have added that now.

**Review by monstermunchkin — APPROVED** [review]

File: lxd/db/warnings.go [hunk 1]
```diff
@@ -51,6 +51,31 @@ SELECT warnings.id FROM warnings LEFT JOIN nodes ON warnings.node_id = nodes.id
...
+
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "Failed getting local member name")
+	}
+
+	if localName == "" {
+		return fmt.Errorf("Local member name not available")
+	}
```
Comment (monstermunchkin) [inline lxd/db/warnings.go:73]: @tomponline What's the reason we fail here? When running non-clustered LXD the node name is empty and shouldn't fail here. Or am I missing something?

Comment (tomponline) [reply lxd/db/warnings.go:73]: In a standalone lxd the local node name is 'none' so if its empty something has gone wrong, which is what I saw during my unrelated cluster upgrade tests that caused the heartbeat request to be unnecessarily ended due to not being able to find the node name due to the cluster not being fully started.

Comment (tomponline) [reply lxd/db/warnings.go:73]: As the warnings themselves are stored using the node id we could go further still and remove the indirection and modify the functions to take nodr id directly instead which would also reduce the amount of queries.

Comment (monstermunchkin) [reply lxd/db/warnings.go:73]: You're right, it's "none", not empty. I missread my db output. Sorry for the noise ;)

**Pull request merged**